### PR TITLE
Unlimited message cache

### PR DESCRIPTION
**Pull request by cmebarrow**
Requires https://github.com/reaktivity/nukleus-kafka.spec/pull/61 (to fix an occasional build failure, unrelated to the other changes in this PR)

Added support for arbitrarily large message cache. MemoryLayout assigns multiple memory buffers if needed,  all buffers except the last being of size 1 GB. There is still only one metadata tree and metadata buffer. All buffers are mapped to the same file.

Also added counters for historical fetches, one per Kafka broker network. Example output:
```
java -Dreaktor.streams.buffer.capacity=0x2000000 -Dreaktor.throttle.buffer.capacity=0x200000 -jar command-log.jar -t counters -d shm | tee counters.txt |grep kafka

kafka-sse.routes 1
kafka-sse.acquires 0
kafka-sse.releases 0
kafka.routes 1
kafka.acquires 0
kafka.releases 0
kafka.message.cache.buffer.acquires 1260602
kafka.historical.fetches.tcp.1 30223
```

### CODE REVIEW

File: src/main/java/org/reaktivity/nukleus/kafka/internal/stream/NetworkConnectionPool.java [hunk 1]
```diff
@@ -131,6 +133,11 @@
     private static final int KAFKA_SERVER_DEFAULT_DELETE_RETENTION_MS = 86400000;
     private static final PartitionIndex DEFAULT_PARTITION_INDEX = new DefaultPartitionIndex();
 
+    private static final LongSupplier NO_COUNTER = () ->
+    {
+        return 0L;
+    };
+
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/kafka/internal/stream/NetworkConnectionPool.java]: `private static final LongSupplier NO_COUNTER = Long.valueOf(0L)::longValue;`

File: .travis.yml [hunk 2]
```diff
@@ -2,6 +2,11 @@ dist: trusty
 language: java
 before_install:
   - rm ~/.m2/settings.xml
+  - git clone https://github.com/cmebarrow/nukleus-kafka.spec
+  - cd nukleus-kafka.spec
+  - git checkout fix_build_failure
+  - mvn clean install -DskipTests
+  - cd ..
```
Comment (jfallows) [inline .travis.yml]: Are there any spec changes required?

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java [hunk 3]
```diff
@@ -32,39 +32,56 @@
...
         return memoryBuffer.addressOffset() + address;
     }
 
     @Override
     public long acquire(
-        int capacity)
+        final int capacity)
     {
-        if (capacity > this.maximumBlockSize)
+        if (capacity > this.firstBufferCapacity)
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java]: Since `maximumBlockSize` is already guarded by implementation detail of `memoryBuffer.capacity`, shouldn't this remain as `if (capacity > this.maximumBlockSize) ...` ?

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java [hunk 4]
```diff
@@ -16,7 +16,7 @@
 package org.reaktivity.nukleus.kafka.internal.memory;
 
 import static java.lang.Integer.highestOneBit;
-import static java.lang.Integer.numberOfTrailingZeros;
+import static java.lang.Long.numberOfTrailingZeros;
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java]: Also need to change import to `Long.highestOneBit` ?

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java [hunk 5]
```diff
@@ -127,7 +145,8 @@ public long acquire(
             }
         }
 
-        return ((nodeIndex + 1) & ~highestOneBit(nodeIndex + 1)) << blockSizeShift << nodeOrder;
+        long addressToShift = ((nodeIndex + 1) & ~highestOneBit(nodeIndex + 1));
+        return addressToShift << blockSizeShift << nodeOrder;
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java]: Looks like this was a no-op refactor except to promote to `long`, but using `Long.highestOneBit(...)` instead of `Integer.highestOneBit(...)` makes the return value type `long` already.

Please verify that behavior remains unchanged when using `Long.highestOneBit(...)` instead and if so, then eliminate unnecessary diff.

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/MemoryLayout.java [hunk 6]
```diff
@@ -128,67 +125,121 @@ public Builder minimumBlockSize(
...
             if (!isPowerOfTwo(maximumBlockSize))
             {
                 throw new IllegalArgumentException("maximum block size MUST be a power of 2");
             }
 
+            if (maximumBlockSize >> Integer.numberOfTrailingZeros(ONE_GB) > Integer.MAX_VALUE)
+            {
+                throw new IllegalStateException("capacity too large, number of 1GB buffers would exceed Integer.MAX_VALUE");
+            }
+
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/kafka/internal/memory/MemoryLayout.java]: `maximumBlockSize` also cannot be greater than `1GB` due to implementation constraints, right?

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/MemoryLayout.java [hunk 7]
```diff
@@ -42,63 +41,61 @@
...
+        return metadataBuffer.getLong(MAXIMUM_BLOCK_SIZE_OFFSET);
     }
 
     public static final class Builder extends Layout.Builder<MemoryLayout>
     {
         private Path path;
-        private int capacity;
         private int minimumBlockSize;
-        private int maximumBlockSize;
+        private long maximumBlockSize;
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/kafka/internal/memory/MemoryLayout.java]: `maximumBlockSize` and `capacity` are independent concepts, where `maximumBlockSize <= capacity` in general, and in this new implementation `maximumBlockSize <= 1GB` and `capacity <= Integer.MAX_VALUE * 1GB`.

The initial implementation just made a simplifying assumption that `maximumBlockSize == capacity`, but that is not generally accurate. Some of the changes in this PR are locking in that assumption, but it seems like a good opportunity to properly split the concepts, since `capacity` can be greater than `1GB` but `maximumBlockSize <= 1GB` now.

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java [hunk 8]
```diff
@@ -32,39 +32,56 @@
...
+        {
+            memoryBuffer = memoryBuffers[0];
+        }
+        else
+        {
+            int index = (int) (address >> addressShift);
+            memoryBuffer = memoryBuffers[index];
+            address &= addressMask;
+        }
+
```
[the remark this answers is not in the document]
Comment (cmebarrow) [reply src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java]: yes, done

File: src/main/java/org/reaktivity/nukleus/kafka/internal/stream/NetworkConnectionPool.java (hunk 1 above)
Comment (cmebarrow) [reply src/main/java/org/reaktivity/nukleus/kafka/internal/stream/NetworkConnectionPool.java]: done

File: .travis.yml (hunk 2 above)
Comment (cmebarrow) [reply .travis.yml]: Just for a sporadic test failure that caused the Travis build for the PR to fail.   Unrelated to the changes to increase message cache size,  but I just wanted to make sure the build  is as stable as possible.

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java (hunk 3 above)
Comment (cmebarrow) [reply src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java]: The meaning of maximumBlockSize in MemoryLayout is unchanged: it is the overall size of the cache.  I  decided not to try to make a distinction between capacity and maximumBlockSize,  because the information is already conveyed in the array of MemoryBuffers returned by MemoryLayout.getMemoryBuffers().

I could rename `maximumBlockSize` to `capacity` in MemoryLayout  if you think that would clarify matters.  The only downside I can see is it might then be confused with the capacity parameter to MemoryManager.acquire.

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java (hunk 4 above)
Comment (cmebarrow) [reply src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java]: Correct, done

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java (hunk 5 above)
Comment (cmebarrow) [reply src/main/java/org/reaktivity/nukleus/kafka/internal/memory/DefaultMemoryManager.java]: Verified and fixed

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/MemoryLayout.java (hunk 6 above)
Comment (cmebarrow) [reply src/main/java/org/reaktivity/nukleus/kafka/internal/memory/MemoryLayout.java]: The maximum capacity that can be acquired is 1GB, because of the current implementation constraint.  As explained above I decided to preserve the " total memory" (or  overall capacity) meaning of maximumBlockSize.  Basically in the current code it has a double function, so I had to pick one.

File: src/main/java/org/reaktivity/nukleus/kafka/internal/memory/MemoryLayout.java (hunk 7 above)
Comment (cmebarrow) [reply src/main/java/org/reaktivity/nukleus/kafka/internal/memory/MemoryLayout.java]: I was assuming that one day the two concepts will be unified again. But I  would be happy to rename maximumBlockSize to capacity to make  the current meaning clearer.

**Comment by cmebarrow** [comment]
@jfallows Feedback applied
